lxd/images.go

```diff
@@ -1115,93 +1115,226 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
+		var err error
+
+		images, err := tx.GetImages(db.ImageFilter{AutoUpdate: true})
 		if err != nil {
 			return err
 		}
 
-		for _, project := range projects {
-			projectNames = append(projectNames, project.Name)
+		for _, image := range images {
+			imageMap[image.Fingerprint] = append(imageMap[image.Fingerprint], image)
 		}
 
 		return nil
 	})
 	if err != nil {
-		return errors.Wrap(err, "Unable to retrieve project names")
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Error("Error getting cluster members for image auto-update", log.Ctx{"fingerprint": fingerprint, "err": err})
+			continue
+		}
+
+		if len(nodes) == 0 {
+			continue
+		}
+
+		if len(nodes) > 1 {
+			var nodeIDs []int64
+
+			for _, node := range nodes {
+				err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+					var err error
+
+					nodeInfo, err := tx.GetNodeByAddress(node)
+					if err != nil {
+						return err
+					}
+
+					nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+					return nil
+				})
+				if err != nil {
+					logger.Error("Unable to retrieve cluster member information for image update", log.Ctx{"err": err})
+					continue
+				}
+			}
+
+			// If multiple nodes have the image, select one to deal with it.
+			if len(nodeIDs) > 1 {
+				selectedNode, err := shared.GetStableRandomInt64FromList(len(images), nodeIDs)
+				if err != nil {
+					logger.Error("Failed to select cluster member for image update", log.Ctx{"err": err})
+					continue
+				}
+
+				// Skip image update if we're not the chosen cluster member.
+				// That way, an image is only updated by a single cluster member.
+				if d.cluster.GetNodeID() != selectedNode {
+					continue
+				}
+			}
+		}
+
+		var deleteIDs []int
+		var newImage *api.Image
+
+		for _, image := range images {
+			_, imageInfo, err := d.cluster.GetImage(image.Project, image.Fingerprint, image.Public)
+			if err != nil {
+				logger.Error("Failed to get image", log.Ctx{"err": err})
+			}
+
+			newInfo, err := autoUpdateImage(ctx, d, nil, image.ID, imageInfo, image.Project)
+			if err != nil {
+				logger.Error("Failed to update image", log.Ctx{"err": err})
+
+				if err == context.Canceled {
+					return nil
+				}
+			} else {
+				deleteIDs = append(deleteIDs, image.ID)
+			}
+
+			if newImage == nil {
+				newImage = newInfo
+			}
+		}
+
+		if newImage != nil {
+			err := distributeImage(ctx, d, nodes, fingerprint, newImage)
+			if err != nil {
+				logger.Error("Failed to distribute image", log.Ctx{"err": err, "fingerprint": newImage.Fingerprint})
+
+				if err == context.Canceled {
+					return nil
+				}
+			}
+
+			for _, ID := range deleteIDs {
```

Comment:
@monstermunchkin for non-nil errors, should we still try and delete images here?

Reply:
@tomponline Oh yeah, I missed that.

Reply:
@tomponline I had to revert that. It is important that `newInfo` is non-nil. Otherwise the old image will be deleted even if the image has not been updated.

Reply:
@monstermunchkin if distributeImage returns a non-nil error, should we still try to DeleteImage (as it is now) or should we skip it?

Reply:
@tomponline We should still delete the old images.